rclcpp/src/rclcpp/node_interfaces/node_base.cpp

```diff
@@ -31,6 +32,68 @@ using rclcpp::exceptions::throw_from_rcl_error;
 
 using rclcpp::node_interfaces::NodeBase;
 
+namespace
+{
+/// This class bundles together the lifetime of the rcl_node_t handle with the
+/// lifetime of the RCL context. This ensures that the context will remain alive
+/// for as long as the rcl_node_t handle is alive.
+class NodeHandleWithContext
+{
+public:
+  /// The make function returns a std::shared_ptr<rcl_node_t> which is actually
+  /// an alias for a std::shared_ptr<NodeHandleWithContext>. There is no use for
+  /// accessing a NodeHandleWithContext instance directly, because its only job
+  /// is to couple together the lifetime of a context with the lifetime of a
+  /// node handle that depends on it.
+  static std::shared_ptr<rcl_node_t>
+  make(
+    rclcpp::Context::SharedPtr context,
+    std::shared_ptr<std::recursive_mutex> logging_mutex,
+    rcl_node_t * node_handle)
+  {
+    auto aliased_ptr = std::shared_ptr<NodeHandleWithContext>(
+      new NodeHandleWithContext(
+        std::move(context),
+        std::move(logging_mutex),
+        node_handle));
+
+    return std::shared_ptr<rcl_node_t>(std::move(aliased_ptr), node_handle);
+  }
+
+  ~NodeHandleWithContext()
```

Comment:
Can you `= delete` copy assignment/constructors and `= default` move assignment/constructors  to comply with the rule of 5?
(it doesn't seem safe to have a copy constructor/assignment in this case)

Reply:
The `NodeHandleWithContext` class should be inaccessible in general since it's hidden behind an aliased `std::shared_ptr`, but you're right that we should delete them anyway to make the intention of the class absolutely clear. Done here: c454258